tun/tun_illumos.go

```diff
@@ -0,0 +1,355 @@
+		fd, err := unix.Socket(proto, unix.SOCK_DGRAM, 0)
+		if err != nil {
+			return fmt.Errorf("could not open Socket of type %v", proto)
+		}
+
+		// illumos defines the ioctl number as a signed int, but the
+		// common x/sys/unix functions use uint. Hand-cast things to help
+		// the compiler figure it out.
+		reqnum := int(unix.SIOCGLIFMUXID)
+		if err := unix.IoctlLifreq(fd, uint(reqnum), &l); err != nil {
+			continue
+		}
+
+		id := l.GetLifruInt()
+
+		if err := unix.IoctlSetInt(fd, unix.I_PUNLINK, id); err != nil {
+			continue
+		}
+
+		unix.Close(fd)
+	}
+	return nil
+}
+
+func CreateTUN(name string, mtu int) (Device, error) {
+	if name != "tun" {
+		return nil, fmt.Errorf("Interface name must be 'tun'")
+	}
+
+	fd, err := unix.Open(tunNode, unix.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("Could not open TUN (%s)", tunNode)
+	}
+
+	// Ask the TUN driver for a new PPA number:
+	var ppa int
+	var s unix.Strioctl
+	s.Cmd = unix.TUNNEWPPA
+	for try := 0; try < 128; try++ {
+		s.SetInt(try)
+		ppa, err = unix.IoctlSetStrioctlRetInt(fd, unix.I_STR, &s)
+		if err != nil {
+			if err == unix.EEXIST {
+				// PPA already in use, try the next one.
+				continue
+			}
+			unix.Close(fd)
+			return nil, err
+		}
+		break
+	}
+
+	assignedName := fmt.Sprintf("tun%d", ppa)
+	// Immediately do a full cleanup of anything stale left behind by a previous crash
+	tunDestroy(assignedName)
+
+	dgramFD, err := unix.Socket(unix.AF_INET, unix.SOCK_DGRAM, 0)
+	if err != nil {
+		return nil, fmt.Errorf("could not open a UDP Socket")
+	}
+	defer unix.Close(dgramFD)
+
+	// Unclear why, but we need to open a second handle on the tun device for
+	// the ioctls that follow. Reusing the existing one does not work.
+	tunFD, err := unix.Open(tunNode, unix.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("Could not open TUN (%s)", tunNode)
+	}
+	defer unix.Close(tunFD)
+
+	// Push the IP module onto the new TUN device.
+	if err = unix.IoctlSetString(tunFD, unix.I_PUSH, "ip"); err != nil {
+		unix.Close(fd)
+		tunDestroy(assignedName)
+		return nil, fmt.Errorf("unable to push IP module: %v", err)
+	}
+
+	// illumos defines the ioctl number as a signed int, but the
+	// common x/sys/unix functions use uint. Hand-cast things to help
+	// the compiler figure it out.
+	req := int(unix.IF_UNITSEL)
+	err = unix.IoctlSetPointerInt(tunFD, uint(req), ppa)
+	if err != nil {
+		unix.Close(fd)
+		tunDestroy(assignedName)
+		return nil, fmt.Errorf("unable to IF_UNITSEL: %v", err)
+	}
+
+	muxid, err := unix.IoctlSetIntRetInt(dgramFD, unix.I_PLINK, tunFD)
+	if err != nil {
+		unix.Close(fd)
+		tunDestroy(assignedName)
+		return nil, fmt.Errorf("unable to I_PLINK: %v", err)
+	}
+
+	var l unix.Lifreq
+	if err := l.SetName(assignedName); err != nil {
+		unix.Close(fd)
+		tunDestroy(assignedName)
+		return nil, fmt.Errorf("failed to set assignedName on Lifreq :%v", err)
+	}
+
+	// illumos defines the ioctl number as a signed int, but the
```

Comment:
For future readers, instead of these comments on integer casts, it'd be nice to have a comment for each ioctl block explaining what it does, for folks (like me) less familiar with the streams API.